staging/src/k8s.io/client-go/tools/remotecommand/websocket.go

```diff
@@ -0,0 +1,344 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package remotecommand
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"sync"
+	"time"
+
+	gwebsocket "github.com/gorilla/websocket"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/httpstream"
+	"k8s.io/apimachinery/pkg/util/remotecommand"
+	restclient "k8s.io/client-go/rest"
+	"k8s.io/client-go/transport/websocket"
+	"k8s.io/klog/v2"
+)
+
+const (
+	pingReadDeadline = 60 * time.Second
+	// should be less than pingReadDeadline
+	pingWriteDeadline = 10 * time.Second
+	PingPeriod        = 5 * time.Second
+)
+
+var (
+	_ Executor          = &wsStreamExecutor{}
+	_ streamCreator     = &wsStreamCreator{}
+	_ httpstream.Stream = &stream{}
+
+	streamType2streamID = map[string]byte{
+		v1.StreamTypeStdin:  remotecommand.StreamStdIn,
+		v1.StreamTypeStdout: remotecommand.StreamStdOut,
+		v1.StreamTypeStderr: remotecommand.StreamStdErr,
+		v1.StreamTypeError:  remotecommand.StreamErr,
+		v1.StreamTypeResize: remotecommand.StreamResize,
+	}
+)
+
+// wsStreamExecutor handles transporting standard shell streams over an httpstream connection.
+type wsStreamExecutor struct {
+	config    *restclient.Config
+	method    string
+	url       string
+	protocols []string
+}
+
+// NewWebSocketExecutor allows to execute commands via a WebSocket connection.
+func NewWebSocketExecutor(config *restclient.Config, method, url string) Executor {
+	return &wsStreamExecutor{
+		config:    config,
+		method:    method,
+		url:       url,
+		protocols: []string{remotecommand.StreamProtocolV5Name},
+	}
+}
+
+// Deprecated: use StreamWithContext instead to avoid possible resource leaks.
+// See https://github.com/kubernetes/kubernetes/pull/103177 for details.
+func (e *wsStreamExecutor) Stream(options StreamOptions) error {
+	return e.StreamWithContext(context.Background(), options)
+}
+
+func (e *wsStreamExecutor) StreamWithContext(ctx context.Context, options StreamOptions) error {
+	req, err := http.NewRequestWithContext(ctx, e.method, e.url, nil)
+	if err != nil {
+		return err
+	}
+	rt, wsRt, err := websocket.RoundTripperFor(e.config)
+	if err != nil {
+		return err
+	}
+	conn, err := websocket.Negotiate(rt, wsRt, req, e.protocols...)
+	if err != nil {
+		return err
+	}
+	defer conn.Close()
+	if conn == nil {
+		panic(fmt.Errorf("websocket connection is nil"))
+	}
+	var streamer streamProtocolHandler
+
+	streamingProto := conn.Subprotocol()
+	klog.V(4).Infof("The subprotocol is %s", streamingProto)
+	if streamingProto != remotecommand.StreamProtocolV5Name {
+		return fmt.Errorf("unsupported streaming protocol: %q", streamingProto)
+	}
+	streamer = newStreamProtocolV4(options)
```

Comment:
it's weird to turn a negotiated v5 protocol into use of a v4 protocol... if the streaming side of v5 is identical to v4 (which I think it is), then represent that as a streamProtocolV5 type that contains/delegates to a v4 instance (like v4 delegates to a v3 instance), and add a protocol switch block here similar to what we do in the spdy client:

```go
	var streamer streamProtocolHandler

	switch protocol {
	case remotecommand.StreamProtocolV5Name:
		streamer = newStreamProtocolV5(options)
  ...
  }

	creator := newWSStreamCreator(conn)
	go creator.run()
  return streamer.stream(creator)
```

Reply:
(and add v5 to the spdy supported protocols list and switch block as well)

Reply:
Fixed.